sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/Offset.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader;
+
+public abstract class Offset {
+    /**
+     * A JSON-serialized representation of an Offset that is
+     * used for saving offsets to the offset log.
+     * Note: We assume that equivalent/equal offsets serialize to
+     * identical JSON strings.
+     *
+     * @return JSON string encoding
+     */
+    public abstract String json();
```

Comment:
Seems what we need is a string representation of offset, why restrict it to json?

Reply:
/cc @marmbrus 

It's my understanding that it's embedded in other JSON, but I'm not sure if that's accurate or whether it's required.

Reply:
@cloud-fan we have been using json as a format in structured streaming since the beginning to make it future-compatible. With this new Datasource APIs we will continue to use it. If we ever need another string format in future (cant imagine why), we can always add another method with a default implementation.